app/src/main/scala/org/alephium/app/ServerUtils.scala

```diff
@@ -212,6 +213,58 @@ class ServerUtils(implicit
     Right(result)
   }
 
+  def buildMultiGroupTransactions(
+      blockFlow: BlockFlow,
+      query: BuildTransaction
+  ): Try[AVector[BuildTransactionResult]] =
+    for {
+      _ <- Either.cond(
+        query.gasAmount.isEmpty,
+        (),
+        badRequest("Explicit Gas Amount is not allowed")
+      )
+      assetOutputRefs <- query.utxos match {
+        case Some(outputRefs) => prepareOutputRefs(outputRefs).left.map(badRequest)
+        case None             => Right(AVector.empty[AssetOutputRef])
+      }
+      _ <- Either.cond(
+        assetOutputRefs.map(_.hint.groupIndex).toSet.size == 1,
+        (),
+        badRequest("Selected UTXOs are not from the same group")
+      )
+      lockPair <- query.getLockPair()
+      _ <- Either.cond(
+        brokerConfig.contains(lockPair._1.groupIndex),
+        (),
+        badRequest(s"This node cannot serve request for Group ${lockPair._1.groupIndex}")
+      )
+      inputSelection <- blockFlow
+        .getUtxoSelectionOrArbitrary(
+          query.targetBlockHash,
+          lockPair._1,
+          assetOutputRefs,
+          apiConfig.defaultUtxosLimit
+        )
+        .left
+        .map(badRequest)
+      outputInfos = prepareOutputInfos(query.destinations)
+      gasPrice    = query.gasPrice.getOrElse(nonCoinbaseMinGasPrice)
+      _ <- getPositiveAlphRemainderOrFail(lockPair._2, inputSelection, outputInfos, gasPrice).left
+        .map(badRequest)
+      unsignedTxs <- blockFlow
+        .buildMultiGroupTransactions(
+          lockPair._1,
+          lockPair._2,
+          inputSelection,
+          outputGroups = AVector.from(outputInfos.groupBy(_.lockupScript.groupIndex).values),
+          gasPrice,
+          AVector.empty[UnsignedTransaction]
+        )
+        .left
+        .map(failed)
+      txs <- validateUnsignedTransactions(unsignedTxs)
+    } yield txs.map(BuildTransactionResult.from)
```

Comment:
This method is structured such that api-user gets as much information as possible about what is `BadRequest` and his fault and what rather an `InternalServerError` as it should succeed when it comes to `blockFlow.buildMultiGroupTransactions` which just recursively builds transactions from available inputs.